src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -604,8 +615,33 @@ private function assertProductNotDuplicate(Order $order, AddProductToOrderComman
         // If we are targeting a specific invoice check that the ID has not been found in the OrderDetail list
         if (!empty($command->getOrderInvoiceId()) && in_array((int) $command->getOrderInvoiceId(), $invoicesContainingProduct)) {
             $orderInvoice = new OrderInvoice($command->getOrderInvoiceId());
-            $invoiceNumber = $orderInvoice->getInvoiceNumberFormatted((int) Configuration::get('PS_LANG_DEFAULT'), $order->id_shop);
+            $invoiceNumber = $orderInvoice->getInvoiceNumberFormatted((int) $this->configuration->get('PS_LANG_DEFAULT'), $order->id_shop);
             throw new DuplicateProductInOrderInvoiceException($invoiceNumber, 'You cannot add this product in this invoice as it is already present');
         }
     }
+
+    /**
+     * Checks whether the discounts based on product quantities should be applied when recomputing order details
+     *
+     * @param AddProductToOrderCommand $command
+     *
+     * @return bool
+     */
+    private function shouldQuantityDiscountsBeApplied(AddProductToOrderCommand $command): bool
+    {
+        /*
+         * The rule is :
+         * If the qtyDiscount is based on Products
+         * and the item added to the order is a combination
+         * we don't apply the specific prices based on the quantity
+         */
+
+        // If the item is not a combination, wo need to go further
+        if (null === $command->getCombinationId()) {
+            return false;
+        }
+
+        // PS_QTY_DISCOUNT_ON_COMBINATION === 1 means quantity discount is based on Combinations, otherwise it's based on Products
+        return 1 === (int) $this->configuration->get('PS_QTY_DISCOUNT_ON_COMBINATION');
```

Comment:
What I still don't understand is that the configuration value itself shouldn't state if the discount is applied or not
In every case when we add something (whether it's a product or a combination) the discount must be applied.

What this configuration is about, from what I understand, is how do I compute the number of products in my cart:
- is it the sum of matching products (regardless of their combination)
- is it strictly the sum of matching combinations?

Reply:
Same answer
I think the rule is : If item added is a combination and PS_QTY_DISCOUNT_ON_COMBINATION = 0 (discounts are based on products) then the specific price is ignored

Ping @MatShir